### ISSUE TITLE
Vitousek et al., Open World Soundness and Collaborative Blame in Gradual Type Systems

### ISSUE THREAD

**Issue by ccshan** [wonks/Talk_Rehearsal_Feedback#1]

**Comment by david-christiansen** [comment]
It is worth considering not putting important text all the way to the bottom of the slide - often the audience can't see the bottom over the heads in front of them.

**Comment by samth** [comment]
Probably worth clarifying that python lists are mutable and thus need proxies.

**Comment by jsiek** [comment]
First slide:
- Use a separate slide before this to introduce gradual typing.
    The title "Soundness in an open world" is distracting.
- The diagram for open world soundness is good.

**Comment by david-christiansen** [comment]
The syntax coloring on the Python code seems to color non-keywords, namely `int`, `list`, `str`, etc, the same as `def`. This is a bit confusing.

**Comment by david-christiansen** [comment]
Perhaps say "that we call the transient design, or just 'transient' for short" to explicitly make the connection.

**Comment by samth** [comment]
Slide 8: don't use `input()`.

**Comment by jsiek** [comment]
Explain the diagram introducing guarded. You went too fast on that slide.

**Comment by jsiek** [comment]
GTLC versus \lambda^*_\to is confusing.

**Comment by david-christiansen** [comment]
Will everyone know the difference between "checks" and "casts"? A compare-contrast slide could be handy.

**Comment by ccshan** [comment]
For slide 10 (Syntax), you'll probably want more transitions to highlight particular parts you're explaining, instead of relying on pointing (which may not be visible).

**Comment by jsiek** [comment]
Need to motivate the white and black diamonds by talking about open world soundness, perhaps by doing some road mapping before we get here.

**Comment by samth** [comment]
Slide 10: you don't need to explain contexts quite so much

**Comment by ccshan** [comment]
Slide 10 (Syntax), should $e e^p$ be $(e e)^p$?

**Comment by jsiek** [comment]
Did you say what "stuck" meant in terms of what kind of behavior it models in Python?
I think you're using stuck in a non-standard way that deserves explanation.

**Comment by ccshan** [comment]
On slide 12, you first spoke the good news, then showed the text "Good news, right?", then repeated the good news in writing, then brought up the issue of blame. I'd show the good news text as you speak the good news; then you won't have to repeat it after "Good news, right?".

**Comment by jsiek** [comment]
(Slide 13) The proxy is too colorful/busy.

**Comment by david-christiansen** [comment]
You can use the `upquote` LaTeX package to get proper vertical single quotes in your verbatim Python code

**Comment by david-christiansen** [comment]
You were on slide 14/19 when your talk hit the 20 minute mark

**Comment by david-christiansen** [comment]
Do you use both `dyn` and `*` to represent the same thing on slide 14?

**Comment by samth** [comment]
The cloud for the blame map is weird.

**Comment by david-christiansen** [comment]
The makeEqChecker cloud on slide 15 has tiny text.

**Comment by david-christiansen** [comment]
Consider a name other than `fn2` - it sounds a bit like "effin' two".

**Comment by david-christiansen** [comment]
You were on slide 16/19 when you hit the hard time limit of 25 minutes.

**Comment by jsiek** [comment]
In the higher-order blame slide, I get lost in the details and don't really understand the general idea.

**Comment by ccshan** [comment]
Spill the beans early. Instead of/before describing the kind of challenges faced by implementors of graduate typing, can you say up front something like "This talk presents Transient, a way to translate gradual typing into Python that (1) eschews proxies so that native C code can access statically typed objects as is; (2) eschews casts except to map heap locations to blame."?

**Comment by vikraman** [comment]
The text fontsize inside the cloud is too small, also red text on blue background is difficult to read.

**Comment by ccshan** [comment]
The blame map may be a case of "Show me your flowcharts and conceal your tables, and I shall continue to be mystified. Show me your tables, and I won’t usually need your flowcharts; they’ll be obvious." Both to save time and to streamline exposition, when you introduce the blame map (i.e., when you turn to restoring blame assignment), can you say something like "The blame map is a map from a heap location to a set of pairs of source location and type. It is extended by ..., consulted by ..., and garbage collected."? (If you did, sorry I missed it.) Then maybe you won't have to say much more.

**Comment by ccshan** [comment]
Computer Modern is especially hard to read in italic (as in theorem statements).

**Comment by jsiek** [comment]
You don't have to literally read the blame theorem. I'd just say something like "upcasts can't be blamed".

**Comment by david-christiansen** [comment]
Slide 18 purports to be a wrap-up slide, but then it asks a new question (performance). This is a bit confusing to follow.

**Comment by ccshan** [comment]
Slide 18 on efficiency: save time and declutter by replacing the bar chart by "3X without blame, ?X with blame, work in progress".

**Comment by jsiek** [comment]
conclusion slide is too wordy

**Comment by ccshan** [comment]
To wrap up, don't say what we want, say what you got.

**Comment by david-christiansen** [comment]
On slide 19, you had taken 34 minutes.

**Comment by cgswords** [comment]
Times
:clock1:
- 5min: End of slide 4
- 10min: Midway through slide 9
- 15min: Reading slide 11
- 20min: Slide 14, talking about heap lookup
- 28min: Getting to Slide 17
- 30min: talking about transient blame and theorems
- 30:30: slide 18
-  33min: slide 19
- 34min: end

Specific notes:
- Slide 2: picture in the middle / larger to start? And larger fonts!
- Personal taste: I don't like serif fonts for talks
- You should better-indicate that the mixed portion is all Module A
- Get a better snake picture? :snake:
- Slide 3 could be broken into 4 slides, and might be better as a result
- C code 'sees' a thing
- Slide 4 into two slides? It'll make things less busy.
- Put the prose up for the difference in approach (eg define transient) to give users a concrete change? You could play with highlighting, emphasis, and moving words in the sentence to indicate the change
- Your titles are all misaligned (ie not centered) :sob:
- Talk through Slides 7-8 using proxies _and_ transient to more clearly demonstrate the difference [maybe]?
- Drop neon green from Slide 9
- Throw some citations onto slide 10 [maybe]?
- _e_s_ vs _e_ is confusing on Slide 10; explain which they are (maybe label by name)
- get a laser pointer for your talk
- It might be worthwhile to introduce the syntax and translation, then talk about _p_ and markings
- Don't use the phrase 'GTLC' without defining it
- It might be nice to pare down the running examples in the talk. Pick one or two, spend time explaining them, and then revisit them when you need an example for a thing.
- Don't use L7 on slide 14
- What if all the examples were in the formalism? Might be easier to understand (might not).
- 'it is a function' -- what is?
- Explicitly adding the checks in on slide 15 will make it a lot easier to follow
- Draw out the blame heap proper for slide 15
- Just... spend a lot less time on blame
- Slide 17 is in the weeds.
- talk about some specific numbers on the results slide, maybe?
- thrown in an 'other things in the paper' slide, **run** through it
- put links on last slide?

**Comment by jsiek** [comment]
please put in slide numbers for your next practice talk

**Comment by cgswords** [comment]
@jsiek they exist, just hard to see

**Comment by jsiek** [comment]
@cgswords Ahh, Kyle's head was obscuring them!

**Comment by ccshan** [comment]
Consider making notation easier to read and more uniform in the slides, even if it makes the notation more different from the paper. For example, you could replace $\lambda^\dots_\dots$ by English names. More dramatically, you could try to unify the Python-like language you use in examples and the math-like language you use in theorems. And maybe change the black and white diamonds to something like "outside" and "inside".

**Comment by jsiek** [comment]
"no overhead in statically typed code with proxy/standard approach"
Not true (this is why monotonic references are needed).

**Comment by samth** [comment]
Have numbers for PyPy and for guarded when you actually give the talk.

**Comment by samth** [comment]
You didn't mention `==` issues.

**Comment by jsiek** [comment]
at the beginning, remember to acknowledge your coauthors!

**Comment by jsiek** [comment]
The content needs to be cut down considerably, but it would be good to keep one place with some "depth".

**Comment by mcimini** [comment]
Some of my other comments:
- you said you put "use-site" checks, and I thought of a call-by-need fashion. Here instead you check as soon as you enter the function, whether the argument is used or not. 
- The slide that shows the syntax of lambda \star, had a remark which was too bright green.
- slide 10. you recall cast insertion: you missed the pattern-maching of the function type. (though it might be distracting to show that at that point.)
- after you did examples of blames compared to the guarded semantics, you wanted to state you theorems but you started that discussion in a very skeptic mood.. I felt you were about to tell us that you failed in proving something important.
- for blame.. at some point casts pop up and I did not follow why. I thought we wanted to get rid of them.
- when you discuss performance also discuss what kind of benchmarks and why they are of significant size for you to say "these are promising results"
- last slide, starts with "we'd like... " I kept thinking that this slide was for "future work"
- you did not speak about object identity.
- the last slide simply mentions the gradual guarantee, did you speak about it, and about evolution of code?

Overall, this will be a nice talk at POPL!
